### DISCUSSION TITLE
Rtdetr

### DISCUSSION THREAD

**Discussion by ujjwal-usb** [ultralytics/ultralytics#8257] [Q&A]
I want to train rtdetr on custom dataset, I am at intermediate stage you can say for doing such things, can anyone tell me in basic language and step by step that how can I do this, where can I annotate my dataset and how it will proceed, this help of yours will means a lot to me.

**Comment by glenn-jocher** [comment]
Hello! Training RT-DETR on a custom dataset is a great way to tailor the model to your specific needs. Here's a simplified guide to get you started:

1. **Annotate Your Dataset**: You can annotate your images using tools like [CVAT](https://github.com/openvinotoolkit/cvat) or [LabelImg](https://github.com/tzutalin/labelImg). These tools will help you draw bounding boxes around your objects and save the annotations in a format compatible with YOLOv8, such as YOLO or COCO format.

2. **Prepare Your Dataset**: Once annotated, organize your dataset into two folders: one for images and one for labels. The labels should match the names of the images.

3. **Create a Dataset YAML**: Write a `.yaml` file that specifies the paths to your train and validation datasets, the number of classes, and class names. Here's an example structure for your `dataset.yaml`:
    ```yaml
    train: /path/to/train/images
    val: /path/to/val/images
    nc: number_of_classes
    names: ['class1', 'class2', ...]
    ```

4. **Train Your Model**: With your dataset ready and your YAML file set up, you can start training RT-DETR using the following command:
    ```bash
    yolo train data=dataset.yaml model=rtdetr.yaml epochs=100
    ```
    Adjust the `epochs` as needed for your training.

5. **Monitor Training**: Keep an eye on the training process. YOLOv8 will output logs that show you how the training is progressing.

6. **Validate Your Model**: After training, you can validate your model's performance on the validation set:
    ```bash
    yolo val data=dataset.yaml model=best.pt
    ```
    This will give you metrics like mAP to understand how well your model is performing.

Remember, the key is to have a well-annotated dataset and a properly configured dataset YAML file. Good luck with your training! 🚀 If you need further assistance, feel free to ask.

**Reply by binhere** [reply]
Thanks for your instructions; they have been helpful to me. This is the first time I have worked with RT-DETR model and I have a couple of questions:

1. According to part 1 "Annotate Your Dataset", I understand that training RT-DETR model process is compatible with annotation formats like YOLO or COCO. I wonder how the model automatically knows what format it uses to train? I annotated my custom dataset on Roboflow and it has available format to download, including both "YOLOv8" and "COCO" formats

2. At part 4 "Train Your Model", I I was confused by command: "yolo train --data dataset.yaml --model rtdetr.yaml --epochs 100"
why training RT-DETR model but the command starts with "yolo"? Does "--model rtdetr.yaml" indicate that RT-DETR model will be used?

3. How can I customize training process in both ways command/Python code. I want to leverage 2 GPUs on Kaggle. I also want to save training/validation result as images and adjust image size. And I very appreciate if you offer the detail references about training control in these aspects.

4. How can I predict with RT-DETR model in Python code?

**Reply by glenn-jocher** [reply]
@binhere here's a concise response to your RT-DETR questions:

1. The model detects annotation format automatically based on your dataset directory structure and `dataset.yaml` class definitions. Both YOLO and COCO formats work if paths are correctly specified in the YAML. For Roboflow exports, see the Dataset Format Guide.

2. The `yolo` CLI command works for all Ultralytics models (including RT-DETR). The `model=rtdetr.yaml` specifies RT-DETR architecture. This is standard across our model implementations.

3. For multi-GPU training add `device=0,1`. To save results as images use `save-period=1` and view in `runs/detect/train`. Adjust image size with `imgsz=640`. For Python implementation:

```python
from ultralytics import RTDETR
model = RTDETR('rtdetr.yaml')
model.train(data='dataset.yaml', epochs=100, imgsz=640, device=[0,1], project='custom_train')
```

See the Training Arguments Documentation for detailed parameter controls.

**Reply by binhere** [reply]
Thanks for your instruction, I trained model rtdetr-l.pt but I had some issues:

I followed some examples on [Ultr](https://docs.ultralytics.com/models/rtdetr/)

I trained model rtdetr-l.pt with coco8 data which is in YOLO format:
yolo train data=coco8.yaml model=rtdetr-l.pt epochs=10 batch=32

Here is my result
<img width="634" alt="image" src="https://github.com/user-attachments/assets/b0c3b019-ec9e-4017-aeab-bac371f03f6c" />

I also tried to validate val set with the best model:
yolo val data=coco8.yaml model='runs/detect/train/weights/best.pt'

<img width="618" alt="image" src="https://github.com/user-attachments/assets/ced9d0d6-cf13-41de-9ee9-f990dfb67f95" />

The returned result indicates that the best model cannot detect any object.

1. I wonder where the problem is? 
2. what set in my data is used at "Validating runs/detect/train/weights/best.pt" step in the training process?

Please let me know how to fix it. 

Here is my notebook: https://colab.research.google.com/drive/1xn57iZwt8rrVr9fWioSOM76PRGnNVSvq?usp=sharing

In this notebook, I also tried YOLOv11s model with the same commands but only YOLOv11s worked perfectly

**Reply by glenn-jocher** [reply]
@binhere for RT-DETR training issues with no detections:

1. **Dataset Configuration**: Ensure your `coco8.yaml` explicitly defines validation image paths under the `val:` key (not just `test:`). The validator uses only the `val` set during training - verify these images contain annotated objects.

2. **Annotation Compatibility**: While RT-DETR supports YOLO format, we've found some Roboflow exports require conversion to COCO format for optimal performance. Try exporting as COCO format and update your YAML file accordingly.

3. **Training Duration**: RT-DETR typically needs longer training than YOLO models. For coco8 (small dataset), try 100+ epochs with `patience 50` to prevent early stopping.

For specific debugging of your case, please share your complete `coco8.yaml` and a sample annotation file. The [RT-DETR Training Guide](https://docs.ultralytics.com/models/rtdetr) has additional troubleshooting steps under Common Issues.

**Reply by binhere** [reply]
Thank you for your instructions, I already learned informative knowledge.

Indeed, RT-DETR is compatible with YOLO format. I downloaded my custom dataset on Roboflow with format 'YOLOv8'.

Besides that, I found the way from [here](https://github.com/ultralytics/ultralytics/issues/4504) to solve my issues: 

1. Using Python instead of command `yolo val`:

from ultralytics import RTDETR

model = RTDETR('best.pt')
model.val(data="data.yaml")

2. Using `yolo val` command including 'rtdetr' in the name of the file weight

old_file = os.path.join("runs/detect/train/weights", "best.pt")
new_file = os.path.join("runs/detect/train/weights", "best_rtdetr.pt")
os.rename(old_file, new_file)

!yolo val data=data.yaml model={new_file}

I hope Ultralytics will have some notices like this on RT-DETR documentation, which are very helpful for beginners.
All the best on the development to the Ultralytics team.

**Reply by glenn-jocher** [reply]
Thank you for sharing your solution and documentation suggestion! We'll pass this feedback to our docs team for consideration. For CLI validation, you can also specify the model architecture directly in the YAML config rather than renaming files. The [RT-DETR Training Guide](https://docs.ultralytics.com/models/rtdetr) has more details on model configuration options.

**Reply by binhere** [reply]
I have some more questions about YAML config. Now I want to customize model. I want to change dropout rate and backbone
I followed documents to add `dropout=0.2` in the command. 

`!yolo train data=...  model=...  dropout=0.2`
       
But I do not know how to change backbone of model. 
For example, I tried this code on [Ultralytics document](https://docs.ultralytics.com/modes/train/#usage-examples):

from ultralytics import YOLO
model = YOLO("yolo11n.yaml")

I did not see the downloaded YAML file. But if I change the model name to `yolo11n.pt` which is automatically downloaded. I expect to replace the current backbone by Resnet50. I very much appreciate if you could provide step-by-step instructions

**Reply by glenn-jocher** [reply]
@binhere to customize the YOLO11 backbone and modify dropout, you'll need to create/edit a YAML configuration file. Here's a minimal example for ResNet50 backbone replacement:

1. Create `custom.yaml`:
```yaml
backbone:
  - [Conv, 64, 3, 2]
  - [ResNet50, out_channels=256]
  # Add remaining layers...
```

2. Initialize with:
```python
from ultralytics import YOLO
model = YOLO('yolo11n.yaml').load('yolo11n.pt')  # Load architecture from YAML + weights from PT
```

For full guidance on model configuration files, see the Ultralytics Model Customization Documentation. Dropout can be set directly via CLI arguments when supported by the architecture.
